### ISSUE TITLE
Cannot use (or request) retryable writes with limit=0

### ISSUE THREAD

**Issue by renehauck** [Automattic/mongoose#6610]
**Do you want to request a *feature* or report a *bug*?**
bug
**What is the current behavior?**
I get this error: 
```
 MongoError: Cannot use (or request) retryable writes with limit=0
      at Function.MongoError.create (node_modules/mongodb-core/lib/error.js:45:10)
      at toError (node_modules/mongodb/lib/utils.js:149:22)
      at /home/rene/Projects/mongoTypeEs/node_modules/mongodb/lib/collection.js:1201:39
      at handler (node_modules/mongodb-core/lib/topologies/replset.js:1209:22)
      at /home/rene/Projects/mongoTypeEs/node_modules/mongodb-core/lib/connection/pool.js:544:18
      at process._tickCallback (internal/process/next_tick.js:61:11)
```
My mongo connection string containes for security reasons `retryWrites=true`

**If the current behavior is a bug, please provide the steps to reproduce.**
I just execute "remove" on a existing model instance. I need expecially the method "remove" for the middleware `post("remove")`.

**What is the expected behavior?**
Trigger the middleware without the error :smiley: 

**Please mention your node.js, mongoose and MongoDB version.**
node:v10.1.0
mongo:v3.6.5
mongoose:v5.1.6

**Comment by lineus** [comment]
I believe this is an issue with the underlying driver, here's a script using just the native driver that replicates the issue:

### 6610_native.js
```js
#!/usr/bin/env node
'use strict';

const env = require('/Users/lineus/.env');
const { MongoClient } = require('mongodb');

const sharable = env.ATLASSRV
  .replace(/\/[^:]+:[^@]+@/, '//**:**@')
  .replace(/cl[^:]+mon/, 'cluster.mon');

const pkg = require.resolve('mongodb').replace(/\/[^/]*$/, '/package.json');
const ver = require(pkg).version;

async function run() {
  console.log(`version: ${ver}`);
  console.log(`conn string: ${sharable}`);
  const client = await MongoClient.connect(env.ATLASSRV);
  const db = client.db('gh-6610');
  await db.dropDatabase();
  await db.collection('tests').insert({ name: 'Billy' });
  let found = await db.collection('tests').findOne({});
  console.log(found);
  await db.collection('tests').remove({ _id: found._id })
    .catch((e) => {
      console.error(e.message);
      return client.close();
    });
}

run();
```
### Output:
```
issues: ./6610_native.js
version: 3.0.10
conn string: mongodb+srv://**:**@cluster.mongodb.net/test?retryWrites=true
{ _id: 5b3231da1eaff11ab5f61428, name: 'Billy' }
Cannot use (or request) retryable writes with limit=0
issues:
```

According to [this bug in the native driver queue](https://jira.mongodb.org/browse/NODE-1513) there are changes coming to retryWrites in the next driver release. [The retryWrite spec is here](https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst). 

From reading the spec, it doesn't look like `remove` is going to be supported with retryWrites ( this is an assumption on my part based on the fact that the native driver has deprecated `collection.remove()`, plus it's exclusion from the [list of supported operations](https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#id12) ). I've opened up a bug with the native driver [here](https://jira.mongodb.org/browse/NODE-1534) for clarification.

In the meantime, if you want to continue using retryWrites and have something resembling a post hook on the call, one solution would be to create an instance method:

### 6610.js
```js
#!/usr/bin/env node
'use strict';
const env = require('/Users/lineus/.env');
const mongoose = require('mongoose');
mongoose.connect(env.ATLASSRV);
const conn = mongoose.connection;
const Schema = mongoose.Schema;

const schema = new Schema({
  name: String
});

function postDelHook(err, res) {
  if (err) { return console.error(err); }
  return console.log('ok: ', res.ok);
}

schema.methods.delete = async function() {
  return Test.deleteOne({ _id: this._id })
    .then((res) => { return postDelHook(null, res); })
    .catch((err) => { return postDelHook(err, null); });
};

const Test = mongoose.model('test', schema);

const test = new Test({
  name: 'Billy'
});

async function run() {
  await conn.dropDatabase();
  let doc = await test.save();
  await doc.delete();
  let count = await Test.count();
  console.log(`There are ${count} docs remaining.`);
  return conn.close();
}

run();
```
### Output:
```
issues: ./6610.js
ok:  1
There are 0 docs remaining.
issues:
```

**Comment by lineus** [comment]
one other note, I forgot that you can add a post hook for an instance method.

### 6610.js
```js
#!/usr/bin/env node
'use strict';
const env = require('/Users/lineus/.env');
const mongoose = require('mongoose');
mongoose.connect(env.ATLASSRV);
const conn = mongoose.connection;
const Schema = mongoose.Schema;

const schema = new Schema({
  name: String
});

schema.methods.delete = async function() {
  return Test.deleteOne({ _id: this._id });
};

schema.post('delete', function(res) {
  console.log('calling post delete hook.');
  if (res.n > 0) {
    return true;
  }
  return false;
});

const Test = mongoose.model('test', schema);

const test = new Test({
  name: 'Billy'
});

async function run() {
  await conn.dropDatabase();
  let doc = await test.save();
  let bool = await doc.delete();
  if(bool) {
    let count = await Test.count();
    console.log(`There are ${count} docs remaining.`);
  }
  return conn.close();
}

run();
```
### Output:
```
issues: ./6610.js
calling post delete hook.
There are 0 docs remaining.
issues:
```

**Comment by renehauck** [comment]
Thank you for your help :smiley:

**Comment by lineus** [comment]
@renehauck You're welcome! 

We should probably leave this open until mongoose is updated with whatever version of the driver this fix gets included in ( hopefully the next version 😄 )

**Comment by lineus** [comment]
This has been fixed in the 3.1.1 driver release.
